tests/vmi_gpu_test.go

```diff
@@ -0,0 +1,122 @@
+			gpus := []v1.Gpu{
+				v1.Gpu{
+					Name: gpuName,
+				},
+			}
+			randomVMI.Spec.Domain.Devices.Gpus = gpus
+			vmi, apiErr := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(randomVMI)
+			Expect(apiErr).ToNot(HaveOccurred())
+
+			pod := tests.GetPodByVirtualMachineInstance(vmi, tests.NamespaceTestDefault)
+			Expect(pod.Status.Phase).To(Equal(k8sv1.PodPending))
+			Expect(pod.Status.Conditions[0].Type).To(Equal(k8sv1.PodScheduled))
+			Expect(strings.Contains(pod.Status.Conditions[0].Message, "Insufficient "+gpuName)).To(Equal(true))
+			Expect(pod.Status.Conditions[0].Reason).To(Equal("Unschedulable"))
+		})
+
+		It("Should create a valid VMI and appropriate libvirt domain", func() {
+			nodesList, err := virtClient.CoreV1().Nodes().List(metav1.ListOptions{})
+			var gpuName = ""
+			for _, item := range nodesList.Items {
+				resourceList := item.Status.Allocatable
+				for k, v := range resourceList {
+					if strings.HasPrefix(string(k), "nvidia.com") {
+						if v.Value() >= 1 {
+							gpuName = string(k)
+							break
+						}
+					}
+				}
+			}
+			Expect(gpuName).ToNot(Equal(""))
+			randomVMI := tests.NewRandomVMIWithEphemeralDisk(tests.ContainerDiskFor(tests.ContainerDiskCirros))
+			gpus := []v1.Gpu{
+				v1.Gpu{
+					Name: gpuName,
+				},
+			}
+			randomVMI.Spec.Domain.Devices.Gpus = gpus
+			vmi, apiErr := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(randomVMI)
+			Expect(apiErr).ToNot(HaveOccurred())
+			tests.WaitForSuccessfulVMIStart(vmi)
+			domain, err := tests.GetRunningVirtualMachineInstanceDomainXML(virtClient, vmi)
+			Expect(err).ToNot(HaveOccurred())
+			domSpec := &api.DomainSpec{}
+			Expect(xml.Unmarshal([]byte(domain), domSpec)).To(Succeed())
+
+			readyPod := tests.GetRunningPodByVirtualMachineInstance(vmi, tests.NamespaceTestDefault)
+
+			gpuOutput, err := tests.ExecuteCommandOnPod(
+				virtClient,
+				readyPod,
+				"compute",
+				[]string{"printenv", "GPU_PASSTHROUGH_DEVICES"},
+			)
+			log.Log.Infof("%v", gpuOutput)
+			Expect(err).ToNot(HaveOccurred())
+			addrList := parseDeviceAddress(gpuOutput)
+
+			Expect(len(addrList)).To(Equal(len(domSpec.Devices.HostDevices)))
+			for n, addr := range addrList {
+				Expect(domSpec.Devices.HostDevices[n].Type).To(Equal("pci"))
+				Expect(domSpec.Devices.HostDevices[n].Managed).To(Equal("yes"))
+				Expect(domSpec.Devices.HostDevices[n].Mode).To(Equal("subsystem"))
+				dbsfFields, err := util.ParsePciAddress(addr)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(domSpec.Devices.HostDevices[n].Source.Address.Domain).To(Equal("0x" + dbsfFields[0]))
+				Expect(domSpec.Devices.HostDevices[n].Source.Address.Bus).To(Equal("0x" + dbsfFields[1]))
+				Expect(domSpec.Devices.HostDevices[n].Source.Address.Slot).To(Equal("0x" + dbsfFields[2]))
+				Expect(domSpec.Devices.HostDevices[n].Source.Address.Function).To(Equal("0x" + dbsfFields[3]))
+			}
+		})
```

Comment:
is there any check we could do inside the guest to make sure the device is present there? I understand it's GPU so it's harder to come up with a useful scenario than for e.g. NICs, but at the very least we could check that the PCI device is present inside the guest, and that it has expected vendor ID or some other characteristics. What do you think?

Reply:
Yes I added code to check GPU in guest. Currently, most of the container disk VM images do not have `lspci` command installed on them. I am using cirros image which has a very version `lspci`. We will be able to do more involved check if the VM has `lspci` command installed on it.

Reply:
@visheshtanksale "very version" is there a word missing? I am not sure what's the issue with e.g. using Fedora that should have fresh pciutils package.